### PR TITLE
client/core: check if wallet is synchronized for spend actions

### LINKED ISSUE #2436: disallow sending funds until wallet is 100% synced
Sending funds from an unsynchronized wallet leads to making txns spending already spend outputs, and transactions that just won't ever get mined.  This is confusing to users and hard to diagnose in support.

### PR DESCRIPTION
**Pull request by ukane-philemon**
Closes #2436

### CODE REVIEW

File: client/core/core.go [hunk 1]
```diff
@@ -10419,9 +10431,18 @@ func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uin
...
+	if !found {
+		return nil, fmt.Errorf("no %s wallet", unbip(assetID))
+	}
+
+	sw, is := w.Wallet.(asset.ShieldedWallet)
+	if !is {
+		return nil, fmt.Errorf("%s wallet is not a shielded wallet", unbip(assetID))
+	}
+
+	if !w.synchronized() {
```
Comment (chappjc) [inline client/core/core.go:10444]: Would make sense for `ShieldFunds` and `UnshieldFunds`, possibly even `NewShieldedAddress` to check also, so maybe the `shieldedWallet` can return a bool too.

Comment (ukane-philemon) [reply client/core/core.go:10444]: `ShieldFunds` and `UnshieldFunds` are like balance checks, but I don't think they should be affected?

IMO, we could make it more obvious that wallets are still syncing and a tooltip about what they should not do until it's done. What do you think about this?

Comment (chappjc) [reply client/core/core.go:10444]: > `ShieldFunds` and `UnshieldFunds` are like balance checks, but I don't think they should be affected?

No, these make transactions, sending value in transactions between accounts.  They should be blocked if the wallet is not synchronized.

Comment (ukane-philemon) [reply client/core/core.go:10444]: > No, these make transactions, sending value in transactions between accounts

Oh, let me check.

Comment (ukane-philemon) [reply client/core/core.go]: Updated, but not checking in `NewShieldedAddress` yet unless you really think we should.

File: client/core/core.go [hunk 2]
```diff
@@ -5324,6 +5324,10 @@ func (c *Core) Send(pw []byte, assetID uint32, value uint64, address string, sub
 		return nil, err
 	}
 
+	if !wallet.synchronized() {
+		return nil, fmt.Errorf("%s is still syncing", unbip(assetID))
+	}
```
Comment (chappjc) [inline client/core/core.go:5329]: Only other thought is that we could add peers>0 check to this as well.  See other spots where we directly access the xcWallet fields under mutex lock, or use the state method, which is slightly more work.

Comment (ukane-philemon) [reply client/core/core.go]: Updated. `postRequiredBonds` does not check for peer count atm but checks for `wallet.synchronized()`, do you want us to also check for peer count there?

Comment (chappjc) [reply client/core/core.go]: Let's do it.
I don't like transactions going nowhere

Comment (ukane-philemon) [reply client/core/core.go]: Updated my last commit.

**Review by chappjc — APPROVED** [review]
LGTM, and you appear to have fixed a data race in `confirmRedemption`.

**Comment by ukane-philemon** [comment]
> you appear to have fixed a data race in confirmRedemption.

Good thing. I didn't notice.

File: client/core/trade.go [hunk 3]
```diff
@@ -2903,12 +2903,8 @@ func (t *trackedTrade) confirmRedemption(match *matchTracker) (bool, error) {
 	// In some cases the wallet will need to send a new redeem transaction.
 	toWallet := t.wallets.toWallet
 
-	if toWallet.peerCount < 1 {
-		return false, fmt.Errorf("%s wallet has no peers", unbip(toWallet.AssetID))
-	}
```
Comment (ukane-philemon) [inline client/core/trade.go:2908]: Oh, this. 🌚

**Pull request merged**